src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNet/AspNetMvcIntegration.cs

```diff
@@ -317,16 +317,55 @@ public static Scope CreateScope(object controllerContext)
             try
             {
                 // call the original method, inspecting (but not catching) any unhandled exceptions
-                return instrumentedMethod(asyncControllerActionInvoker, asyncResult);
+                var result = instrumentedMethod(asyncControllerActionInvoker, asyncResult);
+
+                if (scope != null)
+                {
+                    SetStatusCode(scope, httpContext.Response.StatusCode);
+                    scope.Dispose();
+                }
+
+                return result;
             }
             catch (Exception ex)
             {
-                scope?.Span.SetException(ex);
+                if (scope != null)
+                {
+                    scope.Span.SetException(ex);
+
+                    if (httpContext != null)
+                    {
+                        // We don't know how long it'll take for ASP.NET to invoke the callback,
+                        // so we store the real finish time
+                        var now = scope.Span.Context.TraceContext.UtcNow;
+                        httpContext.AddOnRequestCompleted(h => OnRequestCompleted(h, scope, now));
+                    }
+                    else
+                    {
+                        scope.Dispose();
+                    }
+                }
+
                 throw;
             }
-            finally
+        }
+
+        private static void OnRequestCompleted(HttpContext httpContext, Scope scope, DateTimeOffset finishTime)
+        {
+            SetStatusCode(scope, httpContext.Response.StatusCode);
+            scope.Span.Finish(finishTime);
+            scope.Dispose();
+        }
+
+        private static void SetStatusCode(Scope scope, int statusCode)
+        {
+            try
+            {
+                scope?.Span?.SetTag(Tags.HttpStatusCode, statusCode.ToString());
+            }
+            catch (Exception ex)
             {
-                scope?.Dispose();
+                Log.Error(ex, "Error while setting status code");
```

Comment:
Here's another one :) 
```suggestion
                Log.Error(ex, "Error while setting span tag with status code");
```